libminifi/include/Repository.h

```diff
@@ -0,0 +1,294 @@
+			{
+				_directory = value;
+			}
+			logger_->log_info("NiFi FlowFile Repository Directory %s", _directory.c_str());
+			if (configure_->get(Configure::nifi_flowfile_repository_max_storage_size, value))
+			{
+				Property::StringToInt(value, _maxPartitionBytes);
+			}
+			logger_->log_info("NiFi FlowFile Max Partition Bytes %d", _maxPartitionBytes);
+			if (configure_->get(Configure::nifi_flowfile_repository_max_storage_time, value))
+			{
+				TimeUnit unit;
+				if (Property::StringToTime(value, _maxPartitionMillis, unit) &&
+							Property::ConvertTimeUnitToMS(_maxPartitionMillis, unit, _maxPartitionMillis))
+				{
+				}
+			}
+			logger_->log_info("NiFi FlowFile Max Storage Time: [%d] ms", _maxPartitionMillis);
+			leveldb::Options options;
+			options.create_if_missing = true;
+			leveldb::Status status = leveldb::DB::Open(options, _directory.c_str(), &_db);
+			if (status.ok())
+			{
+				logger_->log_info("NiFi FlowFile Repository database open %s success", _directory.c_str());
+			}
+			else
+			{
+				logger_->log_error("NiFi FlowFile Repository database open %s fail", _directory.c_str());
+				return false;
+			}
+		}
+
+		return true;
+	}
+	//! Put
+	virtual bool Put(std::string key, uint8_t *buf, int bufLen)
+	{
+		if (!_enable)
+			return false;
+			
+		// persistent to the DB
+		leveldb::Slice value((const char *) buf, bufLen);
+		leveldb::Status status;
+		status = _db->Put(leveldb::WriteOptions(), key, value);
+		if (status.ok())
+			return true;
+		else
+			return false;
+	}
+	//! Delete
+	virtual bool Delete(std::string key)
+	{
+		if (!_enable)
+			return false;
+		leveldb::Status status;
+		status = _db->Delete(leveldb::WriteOptions(), key);
+		if (status.ok())
+			return true;
+		else
+			return false;
+	}
+	//! Get
+	virtual bool Get(std::string key, std::string &value)
+	{
+		if (!_enable)
+			return false;
+		leveldb::Status status;
+		status = _db->Get(leveldb::ReadOptions(), key, &value);
+		if (status.ok())
+			return true;
+		else
+			return false;
+	}
+	//! destroy
+	void destroy()
+	{
+		if (_db)
+		{
+			delete _db;
+			_db = NULL;
+		}
+	}
+	//! Run function for the thread
+	static void run(Repository *repo);
+	//! Start the repository monitor thread
+	virtual void start();
+	//! Stop the repository monitor thread
+	virtual void stop();
+	//! whether the repo is full
+	virtual bool isFull()
+	{
+		return _repoFull;
+	}
+	//! whether the repo is enable 
+	virtual bool isEnable()
+	{
+		return _enable;
+	}
+
+protected:
+	//! Repo Type
+	RepositoryType _type;
+	//! Mutex for protection
+	std::mutex _mtx;
+	//! repository directory
+	std::string _directory;
+	//! Logger
+	std::shared_ptr<Logger> logger_;
+	//! Configure
+	//! max db entry life time
+	Configure *configure_;
+	int64_t _maxPartitionMillis;
+	//! max db size
+	int64_t _maxPartitionBytes;
+	//! purge period
+	uint64_t _purgePeriod;
+	//! level DB database
+	leveldb::DB* _db;
+	//! thread
+	std::thread *_thread;
```

Comment:
just to keep code persistent with other way to do the std::thread in ThreadSchedulingAgent and make sure that we start the thread not on constructor.